server/pulp/server/controllers/repository.py

```diff
@@ -186,20 +194,39 @@ def has_all_units_downloaded(repository):
     :return: True if completely downloaded
     :rtype: bool
     """
-    types = {}
-    q_set = model.RepositoryContentUnit.objects(repo_id=repository.repo_id)
-    for association in q_set.only('unit_id', 'unit_type_id'):
-        unit_ids = types.setdefault(association.unit_type_id, [])
-        unit_ids.append(association.unit_id)
-    for type_id, unit_ids in types.items():
-        _model = plugin_api.get_unit_model_by_id(type_id)
-        for page in paginate(unit_ids):
-            q_set = _model.objects(id__in=page, downloaded=False)
-            if q_set.count():
-                return False
+    for qs in _get_mongoengine_repo_querysets(str(repository.repo_id)):
+        if qs(downloaded=False).count():
+            return False
     return True
 
 
+def _get_mongoengine_repo_querysets(repo_id, repo_content_unit_q=None):
```

Comment:
My reason for having it a private function is that it skips any non-mongoengine models, which at the time felt like odd behaviour to have in a public function. Having slept on it, I think it's reasonable given the function name, and it makes sense to break things out into the two functions you suggested :+1:.

Also ``filter`` sounds like a good way to filter.